airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultCheckConnectionWorker.java

```diff
@@ -78,28 +79,29 @@ public ConnectorJobOutput run(final StandardCheckConnectionInput input, final Pa
         WorkerUtils.gentleClose(process, 1, TimeUnit.MINUTES);
       }
 
+      final Optional<FailureReason> failureReason = WorkerUtils.getJobFailureReasonFromMessages(OutputType.CHECK_CONNECTION, messagesByType);
       final int exitCode = process.exitValue();
+      LOGGER.debug("Check connection job subprocess finished with exit code {}", exitCode);
+
       final Optional<AirbyteConnectionStatus> status = messagesByType
           .getOrDefault(Type.CONNECTION_STATUS, new ArrayList<>()).stream()
           .map(AirbyteMessage::getConnectionStatus)
           .findFirst();
 
-      if (status.isPresent() && exitCode == 0) {
+      final String defaultErrorMessage = String.format("Error checking connection, status: %s, exit code: %d", status, exitCode);
+
+      if (status.isEmpty()) {
+        LOGGER.error(defaultErrorMessage);
+        return WorkerUtils.getJobOutput(OutputType.CHECK_CONNECTION, failureReason, defaultErrorMessage, true);
+      } else {
         final StandardCheckConnectionOutput output = new StandardCheckConnectionOutput()
             .withStatus(Enums.convertTo(status.get().getStatus(), Status.class))
             .withMessage(status.get().getMessage());
-
-        LOGGER.debug("Check connection job subprocess finished with exit code {}", exitCode);
         LOGGER.debug("Check connection job received output: {}", output);
         LineGobbler.endSection("CHECK");
-        return new ConnectorJobOutput().withOutputType(OutputType.CHECK_CONNECTION).withCheckConnection(output);
-      } else {
-        final String message = String.format("Error checking connection, status: %s, exit code: %d", status, exitCode);
-        LOGGER.error(message);
-
-        return WorkerUtils.getJobFailureOutputOrThrow(OutputType.CHECK_CONNECTION, messagesByType, message);
+        ConnectorJobOutput jobOutput = WorkerUtils.getJobOutput(OutputType.CHECK_CONNECTION, failureReason, defaultErrorMessage, false);
+        return jobOutput.withCheckConnection(output);
```

Comment:
I think it's a little confusing having a `getJobOutput` method that doesn't fully get the job output (you still have to set the actual output on the JobOutput). Perhaps naming it something like `prepareJobOutput` or `buildJobOutput` can be clearer.

Reply:
Actually, we already have the FailureReason at this point (it's ok if it's null) so maybe we don't need that method at all and it can be 

```
return new ConnectorJobOutput()
    .withOutputType(CHECK_CONNECTION)
    .withFailureReason(failureReason)
    .withCheckConnection(output);
```

which doesn't seem too bad. Including things like error messages in that method call creates a bit of indirection imo since it's never going to be used

Reply:
(the same comment goes for discover and spec)